pkg/integrations/v2/controller.go

```diff
@@ -57,64 +57,108 @@ func newController(l log.Logger, cfg controllerConfig, globals Globals) (*contro
 
 // run starts the controller and blocks until ctx is canceled.
 func (c *controller) run(ctx context.Context) {
+	type worker struct {
+		ci     *controlledIntegration
+		stop   context.CancelFunc
+		exited chan struct{}
+	}
+	var (
+		workersMut sync.Mutex
+		workers    = make(map[*controlledIntegration]worker)
```

Comment:
I keep going back and forth between using a mutex and using a sync.Map here. Using a sync.Map would be easier to follow, but using a mutex makes sure that we can prevent workers from deleting themselves for a while.